src/Hazelcast.Net.Linq.Async/Visitors/HzExpressionVisitor.cs

```diff
@@ -32,39 +32,57 @@ public override Expression Visit(Expression? node)
+            return (HzExpressionType)node.NodeType switch
             {
-                case HzExpressionType.Map:
-                    return VisitMap((MapExpression)node);
-                case HzExpressionType.Column:
-                    return VisitColumn((ColumnExpression)node);
-                case HzExpressionType.Select:
-                    return VisitSelect((SelectExpression)node);
-                case HzExpressionType.Projection:
-                    return VisitProjection((ProjectionExpression)node);
-                default:
-                    return base.Visit(node);
-            }
+                HzExpressionType.Map => VisitMap((MapExpression)node),
+                HzExpressionType.Column => VisitColumn((ColumnExpression)node),
+                HzExpressionType.Select => VisitSelect((SelectExpression)node),
+                HzExpressionType.Projection => VisitProjection((ProjectionExpression)node),
+                HzExpressionType.Join => VisitJoin((JoinExpression)node),
+                _ => base.Visit(node),
+            };
+        }
+
+        //internal for test purposes
+        internal virtual Expression VisitJoin(JoinExpression node)
+        {
+            var left = Visit(node.Left);
+            var right = Visit(node.Right);
+            var condtition = Visit(node.JoinCondition);
+
+            if (left != node.Left || right != node.Right || condtition != node.JoinCondition)
+                return new JoinExpression(left, right, condtition, node.Type);
+
+            return node;
         }
 
-        private Expression VisitProjection(ProjectionExpression node)
+        //internal for test purposes
+        internal virtual Expression VisitProjection(ProjectionExpression node)
         {
-            throw new NotImplementedException();
+            var visitedSource = (SelectExpression)Visit(node.Source);
+            var visitedProjector = Visit(node.Projector);
+
+            if (node.Source != visitedSource || node.Projector != visitedProjector)
+                return new ProjectionExpression(visitedSource, visitedProjector, visitedProjector.Type);
+
+            return node;
         }
 
-        private Expression VisitSelect(SelectExpression node)
+        //internal for test purposes
+        internal virtual Expression VisitSelect(SelectExpression node)
         {
             var from = Visit(node.From);
             var where = Visit(node.Where);
             var columns = VisitColumnDefinititions(node.Columns);
 
             if (from != node.From || where != node.Where || columns != node.Columns)
-                return new SelectExpression(node.Alias, columns, from, where, node.Type);
+                return new SelectExpression(node.Alias, node.Type, columns, from, where);
 
             return node;
         }
 
-        private ReadOnlyCollection<ColumnDefinition> VisitColumnDefinititions(ReadOnlyCollection<ColumnDefinition> columns)
+        //internal for test purposes
```

Comment:
but it's `protected`?

Reply:
Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/732/commits/cea4111d7f8995ee8ea7fad26026d61c33d9d206